applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetDofListKuttaElement(DofsVectorType& rElementalDofList) const
+{
+    const auto& r_geometry = this->GetGeometry();
+    // Kutta elements have only negative part
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (!r_geometry[i].GetValue(TRAILING_EDGE))
+            rElementalDofList[i] = r_geometry[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[i] = r_geometry[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetDofListWakeElement(DofsVectorType& rElementalDofList) const
+{
+    array_1d<double, NumNodes> distances;
+    GetWakeDistances(distances);
+
+    // Positive part
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (distances[i] > 0)
+            rElementalDofList[i] = GetGeometry()[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[i] = GetGeometry()[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+
+    // Negative part - sign is opposite to the previous case
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (distances[i] < 0)
+            rElementalDofList[NumNodes + i] = GetGeometry()[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[NumNodes + i] =
+                GetGeometry()[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftHandSideNormalElement(
+    MatrixType& rLeftHandSideMatrix, const ProcessInfo& rCurrentProcessInfo)
+{
+    if (rLeftHandSideMatrix.size1() != NumNodes || rLeftHandSideMatrix.size2() != NumNodes)
+        rLeftHandSideMatrix.resize(NumNodes, NumNodes, false);
+    rLeftHandSideMatrix.clear();
+
+    ElementalData<NumNodes, Dim> data;
+
+    // Calculate shape functions
+    GeometryUtils::CalculateGeometryData(GetGeometry(), data.DN_DX, data.N, data.vol);
+
+    const double density = ComputeDensity(rCurrentProcessInfo);
+    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
+
+    // Computing local velocity
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputeVelocity<Dim,NumNodes>(*this);
+    for (unsigned int i = 0; i < Dim; i++){
+        velocity[i] += free_stream_velocity[i];
+    }
+
+    const BoundedVector<double, NumNodes> DNV = prod(data.DN_DX, velocity);
+
+    noalias(rLeftHandSideMatrix) +=
+        data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
+    noalias(rLeftHandSideMatrix) += data.vol * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
+
+    const BoundedMatrix<double, NumNodes, NumNodes> rLaplacianMatrix =
+        data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
```

Comment:
I think this Laplacian matrix is not used. If not, where are you using it?

Reply:
Correct, I forgot to remove these lines last time